pyro/ops/gaussian.py

```diff
@@ -34,28 +35,147 @@ def __init__(self, log_normalizer, info_vec, precision):
+        Evaluate the log density of this Gaussian at a point value::
 
-            `-0.5 * value.T @ precision @ value + value.T @ info_vec + log_normalizer`
+            -0.5 * value.T @ precision @ value + value.T @ info_vec + log_normalizer
 
         This is mainly used for testing.
         """
         if value.size(-1) == 0:
             batch_shape = broadcast_shape(value.shape[:-1], self.batch_shape)
             return self.log_normalizer.expand(batch_shape)
-        result = (-0.5) * torch.matmul(self.precision, value.unsqueeze(-1)).squeeze(-1)
+        result = (-0.5) * self.precision.matmul(value.unsqueeze(-1)).squeeze(-1)
         result = result + self.info_vec
         result = (value * result).sum(-1)
         return result + self.log_normalizer
 
+    def condition(self, value):
+        """
+        Condition this Gaussian on a trailing subset of its state.
+        This should satisfy::
+
+            x.condition(y).dim() == x.dim() - y.size(-1)
+        """
+        assert isinstance(value, torch.Tensor)
+        assert value.size(-1) <= self.info_vec.size(-1)
+
+        n = self.info_vec.size(-1) - value.size(-1)
+        info_a = self.info_vec[..., :n]
+        info_b = self.info_vec[..., n:]
+        P_aa = self.precision[..., :n, :n]
+        P_ab = self.precision[..., :n, n:]
+        P_bb = self.precision[..., n:, n:]
+        b = value
+
+        info_vec = info_a - P_ab.matmul(b.unsqueeze(-1)).squeeze(-1)
+        precision = P_aa
+        log_normalizer = (self.log_normalizer +
+                          -0.5 * P_bb.matmul(b.unsqueeze(-1)).squeeze(-1).mul(b).sum(-1) +
+                          b.mul(info_b).sum(-1))
+        return Gaussian(log_normalizer, info_vec, precision)
+
+    def event_logsumexp(self):
+        """
+        Integrates out all latent state (i.e. operating on event dimensions).
+        """
+        n = self.info_vec.size(-1)
+        chol_P = self.precision.cholesky()
+        chol_P_u = self.info_vec.unsqueeze(-1).triangular_solve(chol_P, upper=False).solution.squeeze(-1)
+        u_P_u = chol_P_u.pow(2).sum(-1)
+        return (self.log_normalizer + 0.5 * n * math.log(2 * math.pi) + 0.5 * u_P_u -
+                chol_P.diagonal(dim1=-2, dim2=-1).log().sum(-1))
+
+
+def mvn_to_gaussian(mvn):
+    """
+    Convert a MultivaiateNormal distribution to a Gaussian.
+
+    :param ~torch.distributions.MultivariateNormal mvn: A multivariate normal distribution.
+    :return: An equivalent Gaussian object.
+    :rtype: ~pyro.ops.gaussian.Gaussian
+    """
+    assert isinstance(mvn, torch.distributions.MultivariateNormal)
+    n = mvn.loc.size(-1)
+    precision = mvn.precision_matrix
+    info_vec = precision.matmul(mvn.loc.unsqueeze(-1)).squeeze(-1)
+    log_normalizer = (-0.5 * n * math.log(2 * math.pi) +
+                      -0.5 * (info_vec * mvn.loc).sum(-1) +
+                      precision.cholesky().diagonal(dim1=-2, dim2=-1).log().sum(-1))
```

Comment:
nit: because scale_tril is always available in MVN, it is faster to use `- mvn.scale_tril.diagonal(dim1=-2, dim2=-1).log().sum(-1))`